### DISCUSSION TITLE
thread safety of pagmo::population methods

### DISCUSSION THREAD

**Discussion by montanaviking** [esa/pagmo2#550] [General]
I would like to link the evolving values of different islands together so that one islands dv cannot deviate too far from that of its neighbors. To do this, I would the problem of one island to know the champion decision vector (champion dv) of that of its neighbor dynamically during evolution.
My question is:
Is it thread safe to call one island pagmo::population.champion_x() from a different thread? For example, e.g. from the fitness method in another island's problem?
Thanks,
Phil

**Comment by darioizzo** [comment]
That is provided by the migration machanism why not using it?

**Reply by montanaviking** [reply]
Hi Darioizzo,
Thanks for your quick response and thanks for all you and your team's work on Pagmo! I'm making extensive use of it for transistor modeling and later, will apply it to circuit optimization.
I will likely use the migration mechanism instead and this would likely tilt the solution in the right direction. In my case, each island represents the circuit's response at a particular frequency. Unfortunately, the there is some degeneracy in the optimum giving a nonphysical optimum where the circuit parameters, represented by the dv, and circuit response end up changing discontinuously from one frequency to another, yet still satisfying the optimizing conditions. This is an undesired solution and I was hoping to dynamically constrain,via the problem::bounds() method (not the fitness method as inadvertently mentioned previously), the solution so that one frequency point's solution i.e. pagmo::island is bound by its neighbors.
But, I think you're right, using migration to exchange members of the populations between island neighbors would likely pull the solution to the desired solution and lift the degeneracy. Yea, also a valid critique might be that my whole algorithm is fundamentally mathematically unstable and ill-conditioned but lifting the degeneracy like this might yield good solutions? One way to find out.

While I have your attention, I would eventually like to re-implement MPI or some other method to migrate across machines in a cluster. I could transfer the data via TCP and serialization but would need some documentation as per the internals of migration? I don't yet see how these are exposed in the present API but then I'm relatively new to this.
I had seen a paper stating that Pagmo 1x made use of MPI but that is no longer available in Pagmo 2x?
Thanks and best,
Phil

**Comment by darioizzo** [comment]
An MPI island is indeed missing, I think @bluescarni had some old plans on how to port it to pagmo2 ...  are there alternatives though at this point?

**Reply by montanaviking** [reply]
At some point, I'd like to get my oars in the water on this. I have three nice servers and would like to run them all at once against an optimization/modeling problem. While a cluster-usable pagmo::island class would be optimal, I'll just run the populations separately on each server, with different population starting seeds and pick the best champions from each server in the end. Clusters can also be used to run independent problems/experiments as well of course, and that's the most efficient way to use them.
I can look at the Pagmo source and perhaps with help understanding it from you and @bluescarni it shouldn't be that hard to migrate individuals across TCP links - I just need to get access to the internals of migration.
While MPI is nice, I don't see any compelling reason to use it and building a serialization, say using Boost or Cereal APIs + TCP sockets should handle that. Thanks, Phil
